### PR TITLE
tsp arm, generate empty class for emptymodel

### PR DESCRIPTION
**Pull request by XiaofeiCao**
fix https://github.com/Azure/autorest.java/issues/2779

code: https://github.com/Azure/autorest.java/commit/be4de91becac78516a85f6f87eed441a999b60de
test: https://github.com/Azure/autorest.java/commit/bfa99d93fcc24a5f3feac555423ab57ad14aaa2b

changes to `update` are expected.

fixed `package-info` in https://github.com/Azure/autorest.java/pull/2859/commits/e59d78744300114f9985d5d235982fdc8c9f0329

### CODE REVIEW

File: typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResource.java [hunk 1]
```diff
@@ -130,7 +130,7 @@ interface WithProperties {
     /**
      * The template for ChildExtensionResource update.
      */
-    interface Update extends UpdateStages.WithProperties {
+    interface Update {
```
Comment (XiaofeiCao) [inline typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResource.java:133]: I need to check why this change.

File: typespec-extension/src/main/java/com/azure/autorest/fluent/TypeSpecFluentPlugin.java [hunk 2]
```diff
@@ -31,7 +32,7 @@ public class TypeSpecFluentPlugin extends FluentGen {
 
     public TypeSpecFluentPlugin(EmitterOptions emitterOptions, boolean sdkIntegration) {
         super(new TypeSpecPlugin.MockConnection(), "dummy", "dummy");
-
+        SchemaUtilAdapter.adaptForTypeSpec();
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/main/java/com/azure/autorest/fluent/TypeSpecFluentPlugin.java:35]: Emm... Given all these, would subclass make more sense (I assume we need subclass for ObjectMapper and ModelMapper)?

Comment (XiaofeiCao) [reply typespec-extension/src/main/java/com/azure/autorest/fluent/TypeSpecFluentPlugin.java:35]: Yeah, I had the same thought before. 
It seemed a bit overkill to have 2 subclasses only to override two identical methods `isPlainObject()`.
But yeah, they are cleaner and less hacky..

File: typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResourceInterfaces.java [hunk 3]
```diff
@@ -40,37 +40,6 @@ Response<ChildExtensionResource> getWithResponse(String resourceUri, String topL
...
-     * @param properties The resource properties to be updated.
-     * @throws IllegalArgumentException thrown if parameters fail the validation.
-     * @throws com.azure.core.management.exception.ManagementException thrown if the request is rejected by server.
-     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return extensionResource of Top Level Arm Resource.
-     */
-    ChildExtensionResource update(String resourceUri, String topLevelArmResourceName, String childExtensionResourceName,
-        Object properties);
-
     /**
```
Comment (XiaofeiCao) [inline typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResourceInterfaces.java:43]: This is expected. As we now have `bodyClientModel` (instead of Object):
https://github.com/Azure/autorest.java/blob/9b59bd39524bd50b79d04458b464d5d4a241680b/fluentgen/src/main/java/com/azure/autorest/fluent/mapper/ResourceParser.java#L235-L247

Thus, the update methods should be in resource model's API, not resource collection's.

File: typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResource.java (hunk 1 above)
Comment (XiaofeiCao) [reply typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResource.java:133]: Finally find the cause:
https://github.com/Azure/autorest.java/blob/9b59bd39524bd50b79d04458b464d5d4a241680b/fluentgen/src/main/java/com/azure/autorest/fluent/mapper/ResourceParser.java#L228-L233

We didn't find `PATCH` operation, and fallback to `PUT`, hence the `properties` be present and of type `ChildExtensionResourceInner`.
Reason we didn't find `PATCH` is because its request body is of type `Object`, not ClientModel. We only count request body of type ClientModel be resource `create` or `update` methods. I doubt there are such cases in Swagger brownfield since I can't imagine `PATCH` request body be {"type": "object"}..

In TypeSpec however, the generated [`ResourceUpdateModel`(xxUpdate)](https://github.com/Azure/typespec-azure/blob/f8ee0803166022ec66cc982222d1090aab10cacb/packages/typespec-azure-resource-manager/lib/foundations/arm.foundations.tsp#L114) will [omit](https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-resource-manager/src/private.decorators.ts#L36-L42) `properties` property if it's empty, making `xxUpdate` has no updatable properties.

File: typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResourceInterfaces.java (hunk 3 above)
Comment (XiaofeiCao) [reply typespec-tests/src/main/java/com/cadl/armresourceprovider/models/ChildExtensionResourceInterfaces.java:43]: Turns out the cause be a little different. 
Since the found method is actually `PUT`, we do have `resourceUpdate`. The `update` methods in the resource collection are not considered resource update methods, hence not be removed. 

The new generated code should be right.

File: typespec-extension/src/main/java/com/azure/autorest/fluent/TypeSpecFluentPlugin.java (hunk 2 above)
Comment (XiaofeiCao) [reply typespec-extension/src/main/java/com/azure/autorest/fluent/TypeSpecFluentPlugin.java]: Updated. Though in fact, we need to subclass 5 classes: ObjectMapper, ModelPropertyMapper, ModelMapper, FluentObjectMapper and FluentModelMapper..

Alternative is something like the combination: we rely on MapperFactory to return different `PlainObjectChecker` implementations. 
When needing the `PlainObjectChecker`, we do something like `Mappers.getPlainObjectChecker().check()`.

File: typespec-extension/src/main/java/com/azure/typespec/mapper/TypeSpecModelMapper.java
Comment (weidongxu-microsoft) [inline typespec-extension/src/main/java/com/azure/typespec/mapper/TypeSpecModelMapper.java:1]: I am fine it is there.

But I thought DPG already be this logic?

Comment (XiaofeiCao) [reply typespec-extension/src/main/java/com/azure/typespec/mapper/TypeSpecModelMapper.java:1]: Right.. I'll try removing the DPG mappers.

Comment (XiaofeiCao) [reply typespec-extension/src/main/java/com/azure/typespec/mapper/TypeSpecModelMapper.java:1]: updated

File: javagen/src/main/java/com/azure/autorest/mapper/ModelPropertyMapper.java [hunk 4]
```diff
@@ -24,12 +24,9 @@
 import java.util.Locale;
 import java.util.stream.Collectors;
 
-public class ModelPropertyMapper implements IMapper<Property, ClientModelProperty> {
+public class ModelPropertyMapper implements IMapper<Property, ClientModelProperty>, NeedsPlainObjectCheck {
     private static final ModelPropertyMapper INSTANCE = new ModelPropertyMapper();
 
-    private ModelPropertyMapper() {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ModelPropertyMapper.java:30]: protected?

**Review by weidongxu-microsoft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/mapper/ModelPropertyMapper.java (hunk 4 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/mapper/ModelPropertyMapper.java:30]: fixed
